File: bfs/numbering.cpp
#include<bits/stdc++.h>
using namespace std;

int board[30][30];
int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};
string a[30];
int b[626];

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int n;
    int count=0;

    cin>>n;
    for(int i=0;i<n;i++) {
        cin>>a[i];
    }
    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            board[i][j] = 0;
        }
    }
    queue<pair<int, int>> Q;
    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            if(a[i][j]=='1'&&board[i][j]==0) {
                int size=0;
                count++;
                board[i][j] = 1;
                Q.push({i,j});
                while(!Q.empty()) {
                    size++;
                    pair<int,int> cur = Q.front();
                    Q.pop();
                    for(int dir=0;dir<4;dir++) {
                        int nx = cur.first + dx[dir];
                        int ny = cur.second + dy[dir];
                        if(nx<0||nx>=n||ny<0||ny>=n) continue;
                        if(board[nx][ny]!=0) continue;
                        if(a[nx][ny]!='1') continue;
                        board[nx][ny]++;
                        Q.push({nx,ny});
                    }
                }
                b[count-1] = size;
            }
        }
    }
    sort(b,b+count);
    cout<<count<<"\n";
    for(int i=0;i<count;i++) {
        cout<<b[i]<<"\n";
    }
    return 0;
}

File: bfs/numbering.py
import sys
input = sys.stdin.readline
from collections import deque

def main():
    n = int(input())
    board = [list(map(int, input().strip())) for _ in range(n)]
    dx = [-1, 0, 1, 0]
    dy = [0, -1, 0, 1]
    
    queue = deque()
    count = 0
    size = []
    
    for i in range(n):
        for j in range(n):
            if board[i][j] == 1:
                queue.append((i, j))
                board[i][j] = 0
                count += 1
                s = 0
                
                while queue:
                    x, y = queue.popleft()
                    s += 1
                    
                    for dir in range(4):
                        nx = x + dx[dir]
                        ny = y + dy[dir]
                        if nx < 0 or nx >= n or ny < 0 or ny >= n:
                            continue
                        if board[nx][ny] != 1:
                            continue
                        
                        queue.append((nx, ny))
                        board[nx][ny] = 0
                
                size.append(s)
    
    print(count)
    size.sort()
    for i in range(len(size)):
        print(size[i])

if __name__ == "__main__":
    main()
